File: ru.kmorozov.library.data.server/src/main/java/ru/kmorozov/library/data/server/storage/ServerStorage.java
package ru.kmorozov.library.data.server.storage;

import org.apache.commons.lang3.StringUtils;
import ru.kmorozov.gbd.core.config.IIndex;
import ru.kmorozov.gbd.core.config.IStorage;
import ru.kmorozov.gbd.core.config.IStoredItem;
import ru.kmorozov.gbd.core.logic.library.LibraryFactory;
import ru.kmorozov.gbd.core.logic.model.book.base.IBookData;
import ru.kmorozov.gbd.core.logic.model.book.base.IPage;
import ru.kmorozov.onedrive.client.OneDriveItem;
import ru.kmorozov.onedrive.client.OneDriveProvider;

import java.io.IOException;
import java.nio.file.Path;
import java.util.HashSet;
import java.util.Set;
import java.util.stream.Stream;

public class ServerStorage implements IStorage {

    private final OneDriveProvider api;

    private final OneDriveItem root;

    private OneDriveItem[] children;

    public ServerStorage(final OneDriveProvider api, final OneDriveItem root) {
        this.api = api;
        this.root = root;
    }

    @Override
    public boolean isValidOrCreate() {
        return this.root != null && this.root.isDirectory();
    }

    @Override
    public IStorage getChildStorage(final IBookData bookData) throws IOException {
        for (final OneDriveItem child : this.getChildren())
            if (child.getName().contains(bookData.getVolumeId()))
                return new ServerStorage(this.api, child);

        String volumeId = bookData.getVolumeId();
        String normalizedName = bookData.getTitle()
                .replace(":", "")
                .replace("<", "")
                .replace(">", "")
                .replace("?", "")
                .replace("/", ".");
        String childName = StringUtils.isEmpty(volumeId) ? normalizedName : normalizedName + ' ' + bookData.getVolumeId();

        return new ServerStorage(this.api, this.api.createFolder(this.root, childName));
    }

    @Override
    public int size() throws IOException {
        return this.api.getChildren(this.root).length;
    }

    @Override
    public Set<String> getBookIdsList() throws IOException {
        Set<String> bookIdsList = new HashSet<>();

        for (final OneDriveItem child : this.getChildren())
            if (child.isDirectory()) {
                String[] nameParts = child.getName().split(" ");
                if (LibraryFactory.isValidId(nameParts[nameParts.length - 1]))
                    bookIdsList.add(nameParts[nameParts.length - 1]);
            }

        return bookIdsList;
    }

    @Override
    public boolean isPageExists(final IPage page) throws IOException {
        for (final OneDriveItem child : this.getChildren())
            if (!child.isDirectory())
                if (child.getName().contains(page.getOrder() + "_" + page.getPid() + "."))
                    return true;

        return false;
    }

    @Override
    public Stream<Path> getFiles() throws IOException {
        return null;
    }

    @Override
    public IStoredItem getStoredItem(final IPage page, final String imgFormat) throws IOException {
        return new ServerStoredItem(this, page, imgFormat);
    }

    @Override
    public void refresh() {
        this.children = null;
    }

    @Override
    public IIndex getIndex(final String indexName, final boolean createIfNotExists) {
        return null;
    }

    private OneDriveItem[] getChildren() throws IOException {
        if (this.children == null)
            this.children = this.api.getChildren(this.root);

        return this.children;
    }

    void deleteItem(final OneDriveItem item) throws IOException {
        this.api.delete(item);
    }

    OneDriveItem saveItem(final ServerStoredItem item) throws IOException {
        final OneDriveItem result = this.isPageExists(item.getPage()) ? this.api.replaceFile(this.root, item.asFile()) : this.api.uploadFile(this.root, item.asFile());

        this.refresh();

        return result;
    }
}

File: ru.kmorozov.library.data.server/src/main/java/ru/kmorozov/library/data/server/storage/ServerStorage.kt
package ru.kmorozov.library.data.server.storage

import org.apache.commons.lang3.StringUtils
import ru.kmorozov.gbd.core.config.IIndex
import ru.kmorozov.gbd.core.config.IStorage
import ru.kmorozov.gbd.core.config.IStoredItem
import ru.kmorozov.gbd.core.logic.library.LibraryFactory
import ru.kmorozov.gbd.core.logic.model.book.base.IBookData
import ru.kmorozov.gbd.core.logic.model.book.base.IBookInfo
import ru.kmorozov.gbd.core.logic.model.book.base.IPage
import ru.kmorozov.onedrive.client.OneDriveItem
import ru.kmorozov.onedrive.client.OneDriveProvider

import java.io.IOException
import java.nio.file.Path
import java.util.HashSet
import java.util.stream.Stream

class ServerStorage(private val api: OneDriveProvider, private val root: OneDriveItem?) : IStorage {

    private var children: Array<OneDriveItem>? = null

    override val isValidOrCreate: Boolean
        get() = root != null && root.isDirectory

    override val bookIdsList: Set<String>
        @Throws(IOException::class)
        get() {
            val bookIdsList = HashSet<String>()

            for (child in getChildren())
                if (child.isDirectory) {
                    val nameParts = child.name.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                    if (LibraryFactory.isValidId(nameParts[nameParts.size - 1]))
                        bookIdsList.add(nameParts[nameParts.size - 1])
                }

            return bookIdsList
        }

    override val items: Stream<IStoredItem>?
        @Throws(IOException::class)
        get() = throw IllegalStateException()

    @Throws(IOException::class)
    override fun getChildStorage(bookData: IBookData): IStorage {
        for (child in getChildren())
            if (child.name.contains(bookData.volumeId!!))
                return ServerStorage(api, child)

        val volumeId = bookData.volumeId
        val normalizedName = bookData.title!!
                .replace(":", "")
                .replace("<", "")
                .replace(">", "")
                .replace("?", "")
                .replace("/", ".")
        val childName = if (StringUtils.isEmpty(volumeId)) normalizedName else normalizedName + ' '.toString() + bookData.volumeId

        return ServerStorage(api, api.createFolder(root!!, childName))
    }

    @Throws(IOException::class)
    override fun size(): Int {
        return api.getChildren(root!!).size
    }

    @Throws(IOException::class)
    override fun isPageExists(page: IPage): Boolean {
        for (child in getChildren())
            if (!child.isDirectory)
                if (child.name.contains(page.order.toString() + "_" + page.pid + "."))
                    return true

        return false
    }

    @Throws(IOException::class)
    override fun getStoredItem(page: IPage, imgFormat: String): IStoredItem {
        return ServerStoredItem(this, page, imgFormat)
    }

    override fun refresh() {
        children = null
    }

    override fun getIndex(indexName: String, createIfNotExists: Boolean): IIndex {
        TODO("Not implemented yet")
    }

    override fun restoreState(bookInfo: IBookInfo) {
        throw IllegalStateException()
    }

    @Throws(IOException::class)
    private fun getChildren(): Array<OneDriveItem> {
        if (children == null)
            children = api.getChildren(root!!)

        return children as Array<OneDriveItem>
    }

    @Throws(IOException::class)
    internal fun deleteItem(item: OneDriveItem) {
        api.delete(item)
    }

    @Throws(IOException::class)
    internal fun saveItem(item: ServerStoredItem): OneDriveItem {
        val result = if (isPageExists(item.page)) api.replaceFile(root!!, item.asFile()) else api.uploadFile(root!!, item.asFile())

        refresh()

        return result
    }
}
